File: day14/advent14.rb
class StateEnum
  include Enumerable
  def initialize(st, init)
    @i = init
    @st = st
  end

  def each
    loop do
      while @i < @st.nums.length do
        @i += 1
        yield @st.nums[@i-1]
      end
      @st.tick
    end
  end
end

class State
  attr_reader :elves, :nums
  def initialize(nums = [3, 7], elves = [0, 1])
    @nums = nums
    @e1, @e2 = elves
  end

  def enum(from = 0)
    StateEnum.new(self, from)
  end

  def make_recipes
    sum = nums[@e1] + nums[@e2]
    nxt = []
    d1 = sum % 10
    sum /= 10
    nums << sum if sum > 0
    nums << d1
  end

  def move_elves
    @e1 = (nums[@e1] + 1 + @e1) % nums.length
    @e2 = (nums[@e2] + 1 + @e2) % nums.length
  end

  def tick
    make_recipes
    move_elves
  end

  def search(patt)
    pi = 0
    enum.each_with_index do |n,i|
      if n == patt[pi]
        return i - pi if pi == patt.length-1
        pi += 1
      else
        pi = 0
      end
    end
  end

  def calc(n)
    e = enum
    e.take n
    (e.take 10).join
  end
end

s = State.new

puts "Part 1: #{s.calc(503761)}"
puts "Part 2: #{s.search([5,0,3,7,6,1])}"

File: day14/advent14.lua
local fun = require("fun")

local function tick(e1, e2, rs)
   local sum = rs[e1+1] + rs[e2+1]
   local d1 = sum % 10
   if (sum // 10) > 0 then
      table.insert(rs, (sum // 10) % 10)
   end
   table.insert(rs, d1)
   return (e1 + rs[e1+1] + 1) % #rs, (e2 + rs[e2+1] + 1) % #rs, rs
end

function iter(st, i)
   if i > #st.rs then
      e1, e2, rs = tick(st.e1, st.e2, st.rs)
      st.e1 = e1
      st.e2 = e2
   end
   return i+1, st.rs[i]
end

local function gen(e1, e2, rs, from)
   return iter, { e1 = e1, e2 = e2, rs = rs }, from or 1
end

print("Part 1: " ..
   table.concat(
      fun.iter(gen(0, 1, { 3, 7 }))
         :drop(503761):take(10):totable()
   )
)

local pat = { 5,0,3,7,6,1 }
local pi = 1
for i, n in gen(0, 1, { 3, 7 }) do
   if pat[pi] == n then
      pi = pi + 1
      if pi > #pat then
         print("Part 2: " .. (i - #pat - 1))
         break
      end
   else
      pi = 1
   end
end
